### DISCUSSION TITLE
Making a single element into an array

### DISCUSSION THREAD

**Discussion by link4531** [TomWright/dasel#141] [Q&A] [answered]
So I am once again asking for for your insight  (please don't ban me, I <3 your Util) and I am once again expecting to discover new realms of my stupidity. ;-)

general process: I have a default/template server.xml and other config files which I have a repeatable process of adding to/modifying based off of XYZ variables related to environment, application, user etc. 

Behold, my slightly familiar server.xml (example)
```>code
<Server>
  <Service>
    <Connector port="123"/>
  </Service>
</Server>
```
I would like to add another connector

```>code
./dasel put string -p xml Server.Service.Connector.[].-port 666
```
but of course 'Connector' is not yet seen as an array yet (makes sense) and I get the response
```>code
Error: exit status 1: Error: could not put value: could not propagate value: selector [type:NEXT_AVAILABLE_INDEX selector:.[].-port] does not support value: [kind:map type:map[string]interface {}] map[-port:123]
```
basically many of the default/template config files I am modifying contain only one element as an example/default and when trying to make changes, I run into problems like this.

Solutions? 
One Alternative is to (from previous dasel discussions) is a 'put document' using something on the lines of:
```>code
dasel put document -p xml -d xml Server.Service '<Connector port="123"/>
<Connector port="666"/>'
```
or my ugly alternative...
```>code
./dasel put string -p xml Server.Service.Connector{Remove this string}.-port 666 | sed 's/{Remove this string//g'  > somefile
```
at which point my previous command starts working as intended... and 'le problem' goes away

- Is there an alternative method that is cleaner than what I am trying?

- Any considerations for allowing a command which contains ...Connector.[] to add an additional 'Connector' even when it is not yet considered an array of Connectors?
- is this "abuse" of dasel, not common place and thus, a feature like this not needed/intended?
- I also understand while this type of change is simple when dealing with a xml structure, the same might not be said if dealing with YAML or JSON as I understand element arrays like this are structured differently. 

p.s.
If needed. I can move my inquiries to another site like stackoverflow if these discussions are not relevant / beneficial to development.

**Comment by TomWright** [comment, marked as answer]
Hey again Link, I'm glad you're getting some use out of dasel!

The issue you're seeing here is because of this: https://daseldocs.tomwright.me/usage/supported-file-types#arrays-lists

More specifically:
> Due to the way that XML is decoded, dasel can only detect something as a list if there are at least 2 items.

Dasel is reading your document as the following (in json form):
```json
{
  "Server": {
    "Service": {
      "Connector": {
        "-port": "123"
      }
    }
  }
}
```

Basically, your breakdown is correct.

In many other data formats you would be expected to define the connectors as an array and this would then work entirely as expected, but since XML doesn't let you define something as an array we run into this issue.

### Is there an alternative method that is cleaner than what I am trying?

I'd basically do what you suggested, but with the caveat that I'd extract value using dasel so you don't need to be aware of what it actually is:

I would probably approach it like this:
Extract the single connector...
```bash
$ CONNECTOR=$(dasel -f file.xml -w json .Server.Service.Connector)
```

Use put document to set connector to an array containing the extracted document...
```bash
dasel put document -f file.xml -d json '.Server.Service.Connector' "[$CONNECTOR, {\"-port\":777}]"
```

### Any considerations for allowing a command which contains ...Connector.[] to add an additional 'Connector' even when it is not yet considered an array of Connectors?

There is the potential to allow dasel to convert a single property into an array when the `.[]` selector is used. This may have to be "hidden" behind another flag to ensure it doesn't break any other uses.

### is this "abuse" of dasel, not common place and thus, a feature like this not needed/intended?

This is the kind of problem I want to solve, but the incompatibility is more based on the differences between file formats rather than the data itself, so it's a hard problem to solve. The potential flag I mentioned above could be a good solution for this though!

### I can move my inquiries to another site like stackoverflow

There's no need for that at all! That's exactly what this Q&A section is for 🙂

**Reply by TomWright** [reply]
Hey @link4531,

If I were to add a flag that would convert a single item to an array when used in conjunction with the append selector, do you have any suggestions for the flag name?

**Reply by link4531** [reply]
I would think -a 
unless you had intent for 'a' already
 ... "convert tree element into an Array" and as you stated this has to be used with  .[] at which point you would also need a response for the inverse .
